app/apis/metro/importer.rb

```diff
@@ -31,28 +31,29 @@ def import!
   end
 
   def delete_old_data!
-    # Order of these matters because dependencies in the data
     # ServiceException and StopTime are child tables that no other tables reference
     # so blowing them away completely is fine. They also don't have remote_ids that
     # would make them easy to match with the source data.
-    ServiceException.where(agency: @agency).delete_all
-    StopTime.where(agency: @agency).delete_all
+    @agency.service_exceptions.where(agency: @agency).delete_all
+    @agency.stop_times.where(agency: @agency).delete_all
 
     # These all have remote_id which makes the easy to identify ones that were
     # removed from the source data. They are also referenced by each other, so
     # we don't want to deal with changing primary keys (plus Stops are currently
     # stored serialized in local storage on the client, so changing IDs would mess
     # those up. Which might be a different problem.)
-    Trip.where(agency: @agency).where("trips.remote_id NOT IN (?)", source.trips.map(&:id)).delete_all
-    Route.where(agency: @agency).where("routes.remote_id NOT IN (?)", source.routes.map(&:id)).delete_all
-    Stop.where(agency: @agency).where("stops.remote_id NOT IN (?)", source.stops.map(&:id)).delete_all
-    Service.where(agency: @agency).where("services.remote_id NOT IN (?)", source.calendars.map(&:service_id)).delete_all
+    @agency.routes.where("routes.remote_id NOT IN (?)", source.routes.map(&:id)).destroy_all
+    @agency.trips.where("trips.remote_id NOT IN (?)", source.trips.map(&:id)).destroy_all
+    @agency.shapes.where("shapes.remote_id NOT IN (?)", source.shapes.map(&:id)).destroy_all
+    @agency.shape_points.joins(:shape).where("shapes.remote_id NOT IN (?)", source.shapes.map(&:id)).destroy_all
+    @agency.stops.where("stops.remote_id NOT IN (?)", source.stops.map(&:id)).destroy_all
+    @agency.services.where("services.remote_id NOT IN (?)", source.calendars.map(&:service_id)).destroy_all
+    GC.start
   end
 
   def import_services!
-    source.each_calendar do |cal|
-      service = Service.find_or_initialize_by(remote_id: cal.service_id, agency: @agency)
-      service.attributes = {
+    source.calendars.each do |cal|
+      Service.find_or_initialize_by(remote_id: cal.service_id, agency: @agency).update!(
```

Comment:
Are we sure that find_or_initialize_by..update! will do an INSERT on a new record?